.github/workflows/ci.yml

```diff
@@ -0,0 +1,110 @@
+# This workflow runs unit tests against the specified python versions on both
+# ubuntu and windows. Additionally, it performs CI against the codebase.
+#
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+
+name: CI
+
+on:
+  push:
+    branches: [ "master" ]
+  pull_request:
+    branches: [ "master" ]
+
+jobs:
+  unit-test:
+    name: Unit Tests
+    
+    runs-on: ${{ matrix.os }}
+    
+    strategy:
+      matrix:
+        python-version: ["3.9", "3.10", "3.11"]
+        os: [ubuntu-latest, windows-latest]
+
+    steps:
+    - uses: actions/checkout@v3
+    
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v3
+      with:
+        python-version: ${{ matrix.python-version }}
+    
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install --upgrade -r requirements.txt
+        pip install -e . 
+    
+    - name: Run Unit Tests
+      run: |
+        coverage run --include=edk2toolext\* -m pytest --junitxml=test.junit.xml --html=pytest_report.html --self-contained-html
+        
+
+    - name: Upload coverage to codecov
+      uses: codecov/codecov-action@v2
+      with:
+        verbose: false
+    
+    - uses: actions/upload-artifact@v3
+      with:
+        name: Artifacts-${{matrix.os}}-${{matrix.python-version}}
+        path: |
+          pytest_report.html
+          test.junit.xml
+      if: failure()
+  
+  ci:
+    name: CI Tests
+    runs-on: ubuntu-latest
+    env:
+      python-version: "3.11"
```

Comment:
The idea is that the tests should actually be run with most recent Python version supported, right?

It would be nice to do something like get the max Python version from the matrix array and use it here. Not sure what that would look like doing something like defining the values outside the `unit-test` job and then referencing them in the matrix and here.

But, maybe at least a comment around Line 23 that says to bump this value based on the new max would be helpful.

Reply:
Solved. Unfortunately, the strategy matrix does not have access to env variables to do this easily. Instead, I had to create an initial job that runs that the CI tests and Unit Tests depends on. This initial job generates an output that the other two jobs are able to use. The unit tests use the entire list of python versions to generate the matrix. The CI Tests only accesses the first index.

https://github.com/tianocore/edk2-pytool-extensions/pull/358/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR15